lib/promscrape: add Prometheus-compatible service discovery for Nomad

mr-karan opened the pull request:
Add `nomad_sd_configs` support for service discovery.

Closes https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3367

Here's an example file:

```
global:
  scrape_interval: 10s

scrape_configs:
  - job_name: "prom_nomad"
    nomad_sd_configs:
      - server: "localhost:4646"
        tags:
          - metrics
```

This is the result it produces:

![image](https://user-images.githubusercontent.com/5689132/209659929-6f238684-6a2b-4191-9ca9-13b31181e362.png)

The labels are similar to what is exported here: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#nomad_sd_config

hagen1778 commented on lib/promscrape/discovery/nomad/api.go:
```diff
@@ -0,0 +1,178 @@
+// maxWaitTime is duration for Nomad blocking request.
+func maxWaitTime() time.Duration {
+	d := discoveryutils.BlockingClientReadTimeout
+	// Nomad adds random delay up to wait/16, so reduce the timeout in order to keep it below BlockingClientReadTimeout.
+	// See https://developer.hashicorp.com/nomad/api-docs#blocking-queries
+	// The timeout cannot exceed 10 minuntes. See https://developer.hashicorp.com/nomad/api-docs#blocking-queries
```
```suggestion
	// Nomad adds random delay up to wait/16, so reduce the timeout in order to keep it below BlockingClientReadTimeout.
	// See https://developer.hashicorp.com/nomad/api-docs#blocking-queries
	d -= d / 16
	// The timeout cannot exceed 10 minuntes. See https://developer.hashicorp.com/nomad/api-docs#blocking-queries
	
```

hagen1778 commented on lib/promscrape/discovery/nomad/api.go:
```diff
@@ -0,0 +1,178 @@
+		return token.String(), nil
+	}
+	t := os.Getenv("NOMAD_TOKEN")
+	// Allow empty token - it shouls work if ACL is disabled in Nomad.
```
```suggestion
	// Allow empty token - it should work if ACL is disabled in Nomad.
```

hagen1778 commented on lib/promscrape/discovery/nomad/watch.go:
```diff
@@ -0,0 +1,310 @@
+}
+
+// newnomadWatcher creates new watcher and starts background service discovery for Consul.
+func newnomadWatcher(client *discoveryutils.Client, sdc *SDConfig, datacenter, namespace string) *nomadWatcher {
```
```suggestion
func newNomadWatcher(client *discoveryutils.Client, sdc *SDConfig, datacenter, namespace string) *nomadWatcher {
```

hagen1778 commented on lib/promscrape/discovery/nomad/watch.go:
```diff
@@ -0,0 +1,310 @@
+	return sns
+}
+
+// watchForServiceNodesUpdates watches for Consul serviceNode changes for the given serviceName.
```
```suggestion
// watchForServiceNodesUpdates watches for Nomad serviceNode changes for the given serviceName.
```

hagen1778 commented on lib/promscrape/discovery/nomad/watch.go:
```diff
@@ -0,0 +1,310 @@
+	f := func() {
+		data, newIndex, err := getBlockingAPIResponse(nw.client, path, index)
+		if err != nil {
+			logger.Errorf("cannot obtain Consul serviceNodes for serviceName=%q from %q: %s", sw.serviceName, clientAddr, err)
```
```suggestion
			logger.Errorf("cannot obtain Nomad services for serviceName=%q from %q: %s", sw.serviceName, clientAddr, err)
```

hagen1778 commented on lib/promscrape/discovery/nomad/watch.go:
```diff
@@ -0,0 +1,310 @@
+		}
+		sns, err := parseServices(data)
+		if err != nil {
+			logger.Errorf("cannot parse Consul serviceNodes response for serviceName=%q from %q: %s", sw.serviceName, clientAddr, err)
```
```suggestion
			logger.Errorf("cannot parse Nomad services response for serviceName=%q from %q: %s", sw.serviceName, clientAddr, err)
```

hagen1778 commented on lib/promscrape/discovery/nomad/service.go:
```diff
@@ -0,0 +1,92 @@
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promutils"
+)
+
+// getServiceLables returns labels for Nomad services with given cfg.
```
```suggestion
// getServiceLabels returns labels for Nomad services with given cfg.
```

hagen1778 commented on lib/promscrape/discovery/nomad/service.go:
```diff
@@ -0,0 +1,92 @@
+func parseServices(data []byte) ([]Service, error) {
+	var sns []Service
+	if err := json.Unmarshal(data, &sns); err != nil {
+		return nil, fmt.Errorf("cannot unmarshal ServiceNodes from %q: %w", data, err)
```
```suggestion
		return nil, fmt.Errorf("cannot unmarshal Services from %q: %w", data, err)
```

hagen1778 commented on lib/promscrape/discovery/nomad/watch.go:
```diff
@@ -0,0 +1,310 @@
+	stopCh      chan struct{}
+}
+
+// newnomadWatcher creates new watcher and starts background service discovery for Consul.
```
```suggestion
// newNomadWatcher creates new watcher and starts background service discovery for Nomad.
```

hagen1778 commented on lib/promscrape/discovery/nomad/watch.go:
```diff
@@ -0,0 +1,310 @@
+		serviceWatchersStopped.Inc()
+
+		// Do not wait for the watcher goroutine to exit, since this may take for up to maxWaitTime
+		// if it is blocked in Consul API request.
```
```suggestion
		// if it is blocked in Nomad API request.
```

hagen1778 commented on lib/promscrape/discovery/nomad/api.go:
```diff
@@ -0,0 +1,178 @@
+			logger.Errorf("cannot parse X-Nomad-Index header value in response from %q: %s", path, err)
+			return
+		}
+		// Properly handle the returned newIndex according to https://www.consul.io/api-docs/features/blocking#implementation-details
```
Is the index logic the same for Nomad as for Consul?

zekker6 replied:
Yes, according to https://developer.hashicorp.com/nomad/api-docs#blocking-queries

hagen1778 reviewed (commented):
LGTM! Please see my comments

@f41gh7 @zekker6 wdyt?

zekker6 reviewed (approved):
LGTM

f41gh7 reviewed (approved):
LGTM

mr-karan commented:
@hagen1778 Thanks a lot for the review. I've pushed the changes.

hagen1778 reviewed (commented):
Would also be great to add a note to the docs/CHANGELOG.md

hagen1778 commented:
@mr-karan please check the CI build issues. Thanks!

mr-karan commented:
@hagen1778 Added a comment clarifying that indexing logic is same. Nomad docs for indexing behaviour don't exist as of now. Also fixed the build issues and added a small note in docs/CHANGELOG.md

hagen1778 commented:
Looks very good! Thanks for your contribution!
